src/data/hash_map.lean

```diff
@@ -9,6 +9,25 @@ import data.array.lemmas
 import algebra.group
 import data.sigma.basic
 
+/-!
+# Hash maps
+
+Defines a hash map data structure, representing a finite key-value map
+with a value type that may depend on the key type.  The structure
+requires a `nat`-valued hash function.  For a hash map with `n`
+buckets, the hash function is taken modulo-`n` to determine the bucket
+associated to a given key.
+
+## Main definitions
+
+* `hash_map`: constructed with `mk_hash_map`.
+
```

Comment:
Could you add an "implementation details" section that says a little bit about the construction? In particular, they're built around arrays, which have some overhead. The VM will destructively update arrays when there's only a single reference, but Lean's reference counting is pretty opaque and hard to predict.

Reply:
I added some more detail based on my understanding of the implementation.

Reply:
@robertylewis When this structure was written, arrays were implemented in C++ as `parray`, which has fairly good performance even if the array is shared, by strategically deferring updates in some way. I recall there being a push to replace that with plain arrays, which have a much worse performance cliff if you aren't careful to use the value linearly, but I forget if that ever landed.

Reply:
@digama0 It's probably been at least two years since I tried to use `hash_map` or `array` but I remember having major issues last time. Is the doc string here accurate to your understanding?

Reply:
@digama0 No, that change never landed.  Particularly with monads it was very easy to cause non-linearity.